### PR TITLE
【Hackathon 9th No.37】add test_top_k_renorm_probs

### PR DESCRIPTION
**Pull request by Echo-Nie**
> 目前写了两个case，分别针对 `top_k=1` 和 `top_k=vocab_size`，目前这里的 `batchsize = 1` 是因为算子实现里面 `batchsize > 1` 好像不支持？如果未来支持的话可以再对添加一个 `batchsize = 2` 的case

第一次提交如上（当时看到源码中是有`size>1`的，但是因为测试后报错，所以以为有bug，报错原因是：拿`size=1`的逻辑去测的`size>1`，丢失了batch维度），下面是错误原因以及修改总结。

---

## 只测 batch\_size = 1且扩展到 batch\_size > 1 时出错

- 单个样本输入：

  ```python
  batch_size = 1
  vocab_size = 5
  probs = np.random.rand(batch_size, vocab_size).astype("float32")
  top_k = np.array([2], dtype="int64")
  renorm_probs = top_k_renorm_probs(probs_tensor, top_k_tensor)[0].numpy()
  renorm_probs = renorm_probs.reshape(batch_size, vocab_size)
  ```

- size>1 输入：
  ```python
  batch_size = 2
  vocab_size = 5
  probs = np.random.rand(batch_size, vocab_size).astype("float32")
  top_k = np.array([2, 3], dtype="int64")
  renorm_probs = top_k_renorm_probs(probs_tensor, top_k_tensor)[0].numpy()
  ```

* 报错：

```
输入 probs shape: [2, 5]
输入 top_k shape: [2]
算子返回 Paddle Tensor shape: [5] 
算子返回 Tensor dtype: paddle.float32
算子返回 Tensor place: Place(gpu:0)
转 numpy 后 shape: (5,)
转 numpy 后数据:
 [0.         0.5831307  0.         0.         0.41686934]
  算子返回 Paddle Tensor shape: [5] ...出问题了
```

## 原因

  1. `[0]` 取了第一维：
     ```python
     renorm_probs = top_k_renorm_probs(...)[0].numpy()
     ```
`[0]` 只适用于 batch\_size=1；batch\_size>1 时，算子直接返回 `[batch_size, vocab_size]`，取 `[0]` 会导致 shape 丢失 batch 维度 → 报错或逻辑不一致

  2. reshape 也不必要：

     ```python
     renorm_probs = renorm_probs.reshape(batch_size, vocab_size)
     ```
算子已经返回正确 shape `[batch_size, vocab_size]`，强制 reshape 反而容易出错

## 修正

不要取 `[0]`，直接使用算子返回的 Tensor

**统一处理 batch**：

  ```python
  probs_tensor = paddle.to_tensor(probs)
  top_k_tensor = paddle.to_tensor(top_k)
  renorm_probs = top_k_renorm_probs(probs_tensor, top_k_tensor).numpy()  # shape: [batch_size, vocab_size]
  ```

**验证每个 batch**：

  ```python
  for b in range(batch_size):
      # check sum
      self.assertAlmostEqual(renorm_probs[b].sum(), 1.0, places=6)
      # check non-top-k
      top_indices = np.argsort(probs[b])[::-1][:top_k[b]]
      for j in range(vocab_size):
          if j not in top_indices:
              self.assertAlmostEqual(renorm_probs[b, j], 0.0, places=6)
  ```

### CODE REVIEW

**Comment by luotao1** [comment]
@Sunny-bot1

File: tests/operators/test_top_k_renorm_probs.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
+# Copyright (c) 2024 PaddlePaddle Authors. All Rights Reserved.
```
Comment (Sunny-bot1) [inline tests/operators/test_top_k_renorm_probs.py]: 这里改成2025

File: tests/operators/test_top_k_renorm_probs.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+class TestTopKRenormProbs(unittest.TestCase):
+
+    def setUp(self):
+        paddle.set_device("gpu")
+        np.random.seed(42)
+
+    def test_basic_functionality(self):
+        """Test the operator"""
+        batch_size = 1
```
Comment (Sunny-bot1) [inline tests/operators/test_top_k_renorm_probs.py]: 这个算子是支持batch_size>1的，可以增加batch_size>1的case

Comment (Echo-Nie) [reply tests/operators/test_top_k_renorm_probs.py:31]:
```
输入 probs shape: [2, 5]
输入 top_k shape: [2]
算子返回 Paddle Tensor shape: [5] 
算子返回 Tensor dtype: paddle.float32
算子返回 Tensor place: Place(gpu:0)
转 numpy 后 shape: (5,)
转 numpy 后数据:
 [0.         0.5831307  0.         0.         0.41686934]
```
`算子返回 Paddle Tensor shape: [5]` 这部分有问题， 似乎是 [TopKRenormProb](https://github.com/Echo-Nie/FastDeploy/blob/develop/custom_ops/gpu_ops/sample_kernels/top_k_renorm_probs.cu#L32) 的问题？ 稍后我再检查一下

Comment (Echo-Nie) [reply tests/operators/test_top_k_renorm_probs.py]: > 这个算子是支持batch_size>1的，可以增加batch_size>1的case

```
输入 probs shape: [2, 5]
输入 top_k shape: [2]
算子返回 Paddle Tensor shape: [5] 
算子返回 Tensor dtype: paddle.float32
算子返回 Tensor place: Place(gpu:0)
转 numpy 后 shape: (5,)
转 numpy 后数据:
 [0.         0.5831307  0.         0.         0.41686934]
```
`算子返回 Paddle Tensor shape: [5] ` 有问题，似乎是 [TopKRenormProb](https://github.com/PaddlePaddle/FastDeploy/blob/develop/custom_ops/gpu_ops/sample_kernels/top_k_renorm_probs.cu#L32) 的问题？ 稍后我再检查一下。

File: tests/operators/test_top_k_renorm_probs.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        probs = np.random.rand(batch_size, vocab_size).astype("float32")
+        probs /= probs.sum(axis=1, keepdims=True)
+        top_k = np.array([2], dtype="int64")
+
+        probs_tensor = paddle.to_tensor(probs)
+        top_k_tensor = paddle.to_tensor(top_k)
+
+        # Call the operator
+        renorm_probs = top_k_renorm_probs(probs_tensor, top_k_tensor)[0].numpy()
+        renorm_probs = renorm_probs.reshape(batch_size, vocab_size)
```
Comment (Sunny-bot1) [inline tests/operators/test_top_k_renorm_probs.py]: 可能是因为你这里取了[0]，这里不需要取第一维，也不需要reshape，转numpy的操作可以放在assertAlmostEqual之前

File: tests/operators/test_top_k_renorm_probs.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        self.assertAlmostEqual(renorm_probs[0].sum(), 1.0, places=6)
+
+    def test_edge_cases(self):
+        """Test the operator with edge-case top_k values"""
+        probs = np.array([[0.1, 0.3, 0.4, 0.2]], dtype="float32")
+
+        # Case 1: top_k = 1, only the max element remains
+        top_k_tensor = paddle.to_tensor(np.array([1], dtype="int64"))
+        renorm_probs = top_k_renorm_probs(paddle.to_tensor(probs), top_k_tensor)[0].numpy()
+        renorm_probs = renorm_probs.reshape(1, -1)
```
Comment (Sunny-bot1) [inline tests/operators/test_top_k_renorm_probs.py]: 同上

File: tests/operators/test_top_k_renorm_probs.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+        self.assertEqual((renorm_probs > 0).sum(), 1)  # Only one non-zero element
+        self.assertAlmostEqual(renorm_probs.sum(), 1.0, places=6)
+
+        # Case 2: top_k = vocab_size, output should match input
+        top_k_tensor = paddle.to_tensor(np.array([4], dtype="int64"))
+        renorm_probs = top_k_renorm_probs(paddle.to_tensor(probs), top_k_tensor)[0].numpy()
+        renorm_probs = renorm_probs.reshape(1, -1)
```
Comment (Sunny-bot1) [inline tests/operators/test_top_k_renorm_probs.py]: 同上

File: tests/operators/test_top_k_renorm_probs.py (hunk 1 above)
Comment (Echo-Nie) [reply tests/operators/test_top_k_renorm_probs.py]: Done, thx !

File: tests/operators/test_top_k_renorm_probs.py (hunk 4 above)
Comment (Echo-Nie) [reply tests/operators/test_top_k_renorm_probs.py]: Done, thx !

**Comment by Echo-Nie** [comment]
## Fix
- 增加 batch_size=2 和 batch_size=3
- 统一了 `batch` 内检查逻辑，避免重复写 `assert` 
- 之后如果增加 `batch_size` 或 `vocab_size` 测试，只需调用 `_check_output` 即可~
- 

## Output
```
=== Running check ===
Input probs shape: (3, 6)
Input top_k: [1 2 4]
W0914 10:01:09.158907  2341 gpu_resources.cc:114] Please NOTE: device: 0, GPU Compute Capability: 8.0, Driver API Version: 13.0, Runtime API Version: 12.8
Output shape: (3, 6)
Output probs:
 [[0.         1.         0.         0.         0.         0.        ]
 [0.         0.47175145 0.         0.         0.         0.5282486 ]
 [0.44425836 0.11332123 0.         0.         0.16236813 0.28005224]]
Batch 0: sum=1.000000
Batch 0: non-top-k indices [0, 2, 3, 4, 5] should be zero
Batch 1: sum=1.000000
Batch 1: non-top-k indices [0, 2, 3, 4] should be zero
Batch 2: sum=1.000000
Batch 2: non-top-k indices [2, 3] should be zero
.
=== Running check ===
Input probs shape: (2, 5)
Input top_k: [2 3]
Output shape: (2, 5)
Output probs:
 [[0.         0.5649906  0.4350094  0.         0.        ]
 [0.         0.         0.3981753  0.2763285  0.32549617]]
Batch 0: sum=1.000000
Batch 0: non-top-k indices [0, 3, 4] should be zero
Batch 1: sum=1.000000
Batch 1: non-top-k indices [0, 1] should be zero
.
=== Running check ===
Input probs shape: (1, 5)
Input top_k: [2]
Output shape: (1, 5)
Output probs:
 [[0.        0.5649906 0.4350094 0.        0.       ]]
Batch 0: sum=1.000000
Batch 0: non-top-k indices [0, 3, 4] should be zero
.
=== Running check ===
Input probs shape: (1, 4)
Input top_k: [1]
Output shape: (1, 4)
Output probs:
 [[0. 0. 1. 0.]]
Batch 0: sum=1.000000
Batch 0: non-top-k indices [0, 1, 3] should be zero

Edge case top_k=vocab_size
Input probs:
 [[0.1 0.3 0.4 0.2]]
Output probs:
 [[0.1 0.3 0.4 0.2]]
.
----------------------------------------------------------------------
Ran 4 tests in 0.197s

OK
```

**Review by Sunny-bot1 — APPROVED** [review]

**Review by luotao1 — APPROVED** [review]
